From 182a8f27ec246e211759f207932cc51e9a6f00cb Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:28:06 -0400
Subject: [tools] perf/ui/browser: Fix segfault caused by off by one handling END key

Message-id: <1408462094-14194-318-git-send-email-jolsa@redhat.com>
Patchwork-id: 88245
O-Subject: [PATCH RHEL7.1 BZ1131394 317/325] perf ui browser: Fix segfault caused by off by one handling END key
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 48d038fcd09fa231e254965c3b69f8f640c9e62d
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Thu Nov 14 15:30:41 2013 -0300

description
===========
$ perf record ls
$ perf report

Press 'down enter end'

Result:

Program received signal SIGSEGV, Segmentation fault.

The UI browser, used on a argv array would access past the end of the
array on SEEK_END because it wasn't using 'nr_entries - 1', fix it.

BugLink: https://bugzilla.kernel.org/show_bug.cgi?id=59291
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/ui/browser.c b/tools/perf/ui/browser.c
index 3648d4e..cbaa7af 100644
--- a/tools/perf/ui/browser.c
+++ b/tools/perf/ui/browser.c
@@ -569,7 +569,7 @@ void ui_browser__argv_seek(struct ui_browser *browser, off_t offset, int whence)
   browser->top = browser->top + browser->top_idx + offset;
   break;
  case SEEK_END:
-  browser->top = browser->top + browser->nr_entries + offset;
+  browser->top = browser->top + browser->nr_entries - 1 + offset;
   break;
  default:
   return;
-- 
1.7.1